File: compiler/C/preprocessor.c
/**
 * The purpose of this file is to handle and remove compiler directives
 */

#include <stdio.h>
#include <string.h>

#include "include/ArrayList.h"
#include "include/HashMap.h"

#include "include/preprocessor.h"
#include "include/token.h"
#include "include/debug.h"

#include "include/lexer.h"
#include "include/normalizer.h"


// TODO
ArrayList* handleConditionalCompilation(ArrayList* tokens){
    return tokens;
}


ArrayList* includeFile(char* filename, ArrayList* includeDirs, HashMap* includedAlready){
    // we have a file pointer 
    dbg("PERFORMING PREVIOUS COMPILATION STEPS ON INCLUDED CODE...\n");

    ArrayList* tokens = performLexicalAnalysis(filename);
    tokens = performBasicNormalization(tokens);
    tokens = handleCompilerDirectives(tokens, includeDirs, includedAlready);

    dbg("FINISHED PREVIOUS COMPILATIONS STEPS ON INCLUDED CODE!\n");
    dbg("INCLUDED CODE OUTPUT:\n");
    dbg(ArrayList_toString(tokens, Token_toString));
    dbg("\n");

    return tokens;
}


// TODO
ArrayList* handleIncludes(ArrayList* tokens, ArrayList* includeDirs, HashMap* includedAlready){
    // keep track of what was already included to handle circular includes
    HashMap* childIncludedAlready = HashMap_copy(includedAlready);
    
    int tokensLength = ArrayList_length(tokens);
    if (tokensLength == 1){
        fatal_error(((Token*) ArrayList_get(tokens, 0)), "Expected a filename after include\n");
    }

    // remove the include token
    ArrayList_remove(tokens, 0);
    tokensLength--;

    printf("INCLUDING: %s\n", ArrayList_toString(tokens, Token_toString));

    Token* currentToken = (Token*) ArrayList_get(tokens, 0);
    if (!strcmp(currentToken->token, "<")){
        // this is a standard lib
        
    } else {
        char* filename;
        int filenameLength = strlen(currentToken->token);
        if (filenameLength < 3){
            fatal_error(currentToken, "Expected a filename after include\n");
        }
        filename = currentToken->token + 1;
        currentToken->token[filenameLength-1] = '\0';


        FILE* fp;
        // check current directory for this file
        fp = fopen(filename, "r");
        if (fp != NULL){
            if (HashMap_containsKey(includedAlready, &filename)){
                fatal_error(currentToken, "This include causes a circular import.\n");
            }
            HashMap_put(childIncludedAlready, &filename, "0\0");
            fclose(fp);
            return includeFile(filename, includeDirs, childIncludedAlready);
        }

        // check each directory/header in the includes
        int includeDirsLength = ArrayList_length(includeDirs);
        for (int j=0; j<includeDirsLength; j++){
            char filepath[FILENAME_MAX];
            char* filepathPointer = (char*) filepath;
            char* currentIncludeDir = *((char**)ArrayList_get(includeDirs, j));
            int currentIncludeDirLength = strlen(currentIncludeDir);
            if (currentIncludeDir[currentIncludeDirLength] == '/'){
                currentIncludeDir[currentIncludeDirLength] = '\0';
            }
            snprintf(filepath, FILENAME_MAX, "%s/%s", currentIncludeDir, filename);
            printf("Trying to open %s\n", filepath);
            fp = fopen(filepath, "r");
            if (fp != NULL){
                printf("File successfully found!\n");
                if (HashMap_containsKey(includedAlready, &filepathPointer)){
                    fatal_error(currentToken, "This include causes a circular import.\n");
                }
                HashMap_put(childIncludedAlready, &filepathPointer, "0\0");
                fclose(fp);
                return includeFile(filepath, includeDirs,childIncludedAlready);
            }
            printf("File did not exist...\n");
        }

    }


    return ArrayList_malloc(tokens->memberSize);
}

// TODO
ArrayList* handleDefines(ArrayList* tokens){
    int tokensLength = ArrayList_length(tokens);
    for (int i=0; i<tokensLength; i++){

    }

    return tokens;
}


ArrayList* handleCompilerDirectives(ArrayList* tokens, ArrayList* includeDirs, HashMap* includedAlready){
    ArrayList* result = ArrayList_malloc(tokens->memberSize);

    Token* hashtag = stringToToken("#");

    int tokensLength = ArrayList_length(tokens);
    for (int i=0; i<tokensLength; i++){
        Token* currentToken = (Token*) ArrayList_get(tokens, i);

        // the current token is hashtag.
        // next tokens should be the directive
        if (Token_equalsToken(currentToken, hashtag)){
            if (i + 1 >= tokensLength){
                fatal_error(currentToken, "Expected a compiler directive after #.");
            }

            Token* theDirective = (Token*) ArrayList_get(tokens, i+1);
            if (!strcmp(theDirective->token, "define")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "undef")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "include")){
                // skip adding tokens until ;
                ArrayList* includedFile = ArrayList_malloc(sizeof(Token));
                i += 1;
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    ArrayList_append(includedFile, (Token*) ArrayList_get(tokens, i));
                    i++;
                }
                includedFile = handleIncludes(includedFile, includeDirs, includedAlready);
                int includedFileLength = ArrayList_length(includedFile);
                for (int j=0; j<includedFileLength; j++){
                    ArrayList_append(result, (Token*) ArrayList_get(includedFile, j));
                }
                continue;
            } else if (!strcmp(theDirective->token, "pragma")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "ifndef")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "ifdef")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "if")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "else")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "error")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "__FILE__")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "__LINE__")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "__DATE__")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "__TIME__")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else if (!strcmp(theDirective->token, "__TIMESTAMP__")){
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            } else {
                // TODO
                while (i < tokensLength){
                    if (!strcmp(((Token*)ArrayList_get(tokens, i))->token, ";")){
                        break;
                    }
                    i++;
                }
                continue;
            }

        }

        ArrayList_append(result, currentToken);
    }


    return result;
}

File: compiler/python/preprocessor.py

import os

from token import *
from debug import *

import lexer
import normalizer



class Preprocessor:
    def __init__(self, tokens:list[Token], include_dirs:list[str], filename:str):
        self.tokens = tokens
        self.filename = filename

        self.definitions = {} # definition name : list of tokens

        splitted = filename.split("/")
        if len(splitted) <= 1:
            base_filename = ""
        else:
            base_filename = "/".join(splitted[:-1]) + "/"
        self.tokens = self.handle_compiler_directives(self.tokens, include_dirs, set(), base_filename)


    def handle_compiler_directives(self, tokens:list[Token], include_dirs:list[str], included_already:set, base_filename) -> list[Token]:
        dbg("Handling compiler directives...")

        i = 0
        n = len(tokens)

        result = []

        # TODO: replace with full solution
        conditional_stack = []
        conditional_types = []
        while i < n:
            # handle defined words
            if tokens[i].token in self.definitions:
                substitution = self.definitions[tokens[i].token]
                for x in substitution:
                    result.append(x)
                i += 1
                continue

            if tokens[i] == "#":
                i += 1

                directive_tokens = []
                while i < n and tokens[i].token != "#END_DIRECTIVE":
                    directive_tokens.append(tokens[i])
                    i += 1

                if directive_tokens[0] == "define":
                    if len(directive_tokens) == 1:
                        error(directive_tokens[0], "Expected a token after define...")
                    self.definitions[directive_tokens[1].token] = directive_tokens[2:]
                    dbg(f"Added definition! {directive_tokens[1].token} {self.definitions[directive_tokens[1].token]}")
                elif directive_tokens[0] == "undef":
                    if len(directive_tokens) == 1:
                        error(directive_tokens[0], "Expected a token after undef...")
                    if directive_tokens[1].token in self.definitions:
                        self.definitions.pop(directive_tokens[1].token)
                elif directive_tokens[0] == "include":
                    dbg("Including file")
                    new_included_already = included_already.copy()
                    included_file = self.handle_include(directive_tokens, include_dirs, new_included_already, base_filename)
                    for x in included_file:
                        result.append(x)
                elif directive_tokens[0] == "pragma":
                    pass
                elif directive_tokens[0] == "ifndef":
                    if len(directive_tokens) == 1:
                        error(directive_tokens[0], "Expected a token after ifndef...")
                    if directive_tokens[1].token not in self.definitions:
                        # it is defined. look for 
                        conditional_stack.append(True)
                        conditional_types.append("ifndef")
                    else:
                        # it is not defined
                        conditional_stack.append(False)
                        conditional_types.append("ifndef")
                elif directive_tokens[0] == "ifdef":
                    if len(directive_tokens) == 1:
                        error(directive_tokens[0], "Expected a token after ifdef...")
                    if directive_tokens[1].token in self.definitions:
                        # it is defined. look for 
                        conditional_stack.append(True)
                        conditional_types.append("ifdef")
                    else:
                        # it is not defined
                        conditional_stack.append(False)
                        conditional_types.append("ifdef")

                elif directive_tokens[0] == "if":
                    pass
                elif directive_tokens[0] == "elif":
                    pass
                elif directive_tokens[0] == "endif":
                    if len(conditional_stack) == 0:
                        fatal_error(directive_tokens[0], "Unmatched endif...")
                    while len(conditional_stack) > 0:
                        conditional_stack.pop()
                        last_type = conditional_types.pop()
                        if last_type in ["if", "ifdef", "ifndef"]:
                            break
                elif directive_tokens[0] == "else":
                    if len(conditional_stack) == 0:
                        fatal_error(directive_tokens[0], "No if found before else...")
                    if conditional_stack[-1]:
                        conditional_stack.append(False)
                    else:
                        conditional_stack.append(True)
                    conditional_types.append("else")
                elif directive_tokens[0] == "error":
                    pass
                elif directive_tokens[0] == "__FILE__":
                    pass
                elif directive_tokens[0] == "__LINE__":
                    pass
                elif directive_tokens[0] == "__DATE__":
                    pass
                elif directive_tokens[0] == "__TIME__":
                    pass
                elif directive_tokens[0] == "__TIMESTAMP__":
                    pass
                else:
                    fatal_error(directive_tokens[0], "Unrecognized Preprocessor Directive...")

                i += 1
                continue

            if len(conditional_stack) == 0 or conditional_stack[-1]:
                result.append(tokens[i])

            i += 1

        dbg("Finished handling compiler directives.")
        return result


    def handle_include(self, tokens:list[Token], include_dirs:list[str], included_already:set, base_filename:str) -> list[Token]:
        print(tokens)

        dbg("Performing previous compilation steps on included code...")
        if len(tokens) < 2:
            fatal_error(tokens[0], "Expected a filename after include...")

        filename = ""
        if tokens[1].token == "<" and tokens[-1].token == ">":
            filename = tokens[2:-1]
            return self.handle_library_include(filename)
        elif len(tokens[1].token) > 2 and tokens[1].token[0] == '"' and tokens[-1].token[-1] == '"':
            filename = tokens[1].token[1:-1]

        filename = base_filename + filename

        # look for this file name in each include dir
        the_dir = None
        if os.path.isfile(filename):
            the_dir = "./"
        else:
            for x in include_dirs:
                if os.path.isfile(x + filename):
                    the_dir = x
                    break

        if the_dir == None:
            fatal_error(tokens[0], f"Local file {filename} does not exist...")

        filename = f"{the_dir}{filename}"
        while filename[:2] == "./":
            filename = filename[2:]
        dbg(f"Including local file: {filename}")

        if filename in included_already:
            fatal_error(tokens[0], f"This is a circular include...\n{included_already}")

        included_already.add(filename)
        result = lexer.Lexer(filename).tokens
        result = normalizer.Normalizer(result).tokens

        splitted = filename.split("/")
        if len(splitted) <= 1:
            new_base_filename = ""
        else:
            new_base_filename = "/".join(splitted[:-1]) + "/"

        result = self.handle_compiler_directives(result, include_dirs, included_already, new_base_filename)

        dbg("Finished Previous compilation steps on included code!")
        return result
        

    def handle_library_include(self, filename:str)->list[Token]:
        dbg(f"Including library: {filename}")
        return []

        
